[RFC][authenticator] Introduce a new trait TransactionAuthenticator

imsungchoi-fb opened the pull request:
## Motivation
This PR introduces a new trait TransactionAuthenticator and updated the code base accordingly. The purpose of the PR is to increase reusability of diem code for other blockchains, specifically for authentication scheme. 

As of now, Diem is tightly coupled with `TransactionAuthenticator` enum which is maintained as part of diem.  If a different blockchain wants to reuse diem code with an authentication scheme not yet supported by Diem (e.g., ECDSA), the blockchain developer have to contribute to Diem repo to add a new variant of `TransactionAuthenticator` enum. With this PR, we want to 

Changes include

* Introduce a new trait TransactionAuthenticator 
* Change the existing TransactionAuthenticator enum implementation to DiemTransactionAuthenticator
* Make SignedTransaction generic over TransactionAuthenticator
* Add type parameters for DiemTransactionAuthenticator in diem code accordingly 

## Some check/review points 
* Using `DiemTransactionAuthenticator` instead of `TransactionAuthenticator` would be a breaking change? e.g., No issue for ser/des
* How far do we want to push reusability of diem code? i.e., For now, we added `DiemTransactionAuthenticator` to struct `Transaction`

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

zekun000 reviewed (commented):
hmm, I was expecting something like
```
pub enum TransactionAuthenticator {
    /// Single signature
    Ed25519 {
        public_key: vec<u8>,
        signature: vec<u8>,
    },
    /// K-of-N multisignature
    MultiEd25519 {
        public_key: vec<u8>,
        signature: vec<u8>,
    },
```
so the behavior is implementation specific instead of specification specific (the underlying bytes representation stays the same). or in other words, the choice of eager/lazy should be user agnostic.

emmazzz commented:
> hmm, I was expecting something like
> 
> ```
> pub enum TransactionAuthenticator {
>     /// Single signature
>     Ed25519 {
>         public_key: vec<u8>,
>         signature: vec<u8>,
>     },
>     /// K-of-N multisignature
>     MultiEd25519 {
>         public_key: vec<u8>,
>         signature: vec<u8>,
>     },
> ```
> 
> so the behavior is implementation specific instead of specification specific (the underlying bytes representation stays the same). or in other words, the choice of eager/lazy should be user agnostic.

`LazyEvaluation` (what's in the PR currently) is what we have in the design doc. It was called `ToBeDeserialized` in the doc and @imsungchoi-fb gave it a better name. The advantage of adding a new `LazyEvaluation` variant over changing the current `Ed25519` and `MultiEd25519` variants is agility outside of Diem repo. With a more generic adapter interface, other instantiations of Diem can reuse the infrastructure, use `LazyEvaluation` variant to implement their own authentication scheme and signature verification algorithm without having to change any Diem code.

Do you have a strong preference for implementation specific behavior @zekun000 ? Also if we are changing the variants, will there be backwards compatibility issues?

zekun000 commented:
ah, I thought this is for performance so I was expecting we only change implementation. for agility purpose it makes sense, although the name is a little confusing that makes me feel it's for optimization purpose (just my opinion).

we should be able to maintain the compatibility by changing the pubkey/signature <-> vec<u8>, but we probably should do it in another PR.

imsungchoi-fb commented:
> ah, I thought this is for performance so I was expecting we only change implementation. for agility purpose it makes sense, although the name is a little confusing that makes me feel it's for optimization purpose (just my opinion).
> 
> we should be able to maintain the compatibility by changing the pubkey/signature <-> vec, but we probably should do it in another PR.

I've changed the original enum name `ToBeSerialized` to `LazyEvaluation` as I thought it is more accurate for what we are doing. I have no strong opinion, but I am willing to change the name if you guys think that `ToBeSerialized` or another one is making more sense. Feel free to add your input on naming.

zekun000 commented:
I'm not fancy enough, I was thinking about `Others` lol (which shouldn't support existing variants).

also wonder what's the thought for future variants that are not just pubkey and signatures like the multi-agents variant?

imsungchoi-fb commented:
> I'm not fancy enough, I was thinking about `Others` lol (which shouldn't support existing variants).
> 
> also wonder what's the thought for future variants that are not just pubkey and signatures like the multi-agents variant?

lol. I prefer to my suggestions against `Others` and will wait for @emmazzz's input.  Naming is the hardest problem :/ 

For multi-agents, we have the different authenticator `AccountAuthenticator` and we added the same variant to the authenticator as follows. 
In the future, if we have a new scheme to support (e.g., ECDSA), I think we don't have to add a new variant "ECDSA" to both of authenticators. We just need to add "ECDSA" scheme to Scheme and reuse the LazyEvaluation variant. 

@emmazzz Please correct me if I am wrong. 

```rust
pub enum AccountAuthenticator {
    /// Single signature
    Ed25519 {
        public_key: Ed25519PublicKey,
        signature: Ed25519Signature,
    },
    /// K-of-N multisignature
    MultiEd25519 {
        public_key: MultiEd25519PublicKey,
        signature: MultiEd25519Signature,
    },
    /// Delay Public Key & Signature Deserialization until Verify
    LazyEvaluation {
        scheme: u8, // which scheme to deserialize this into
        public_key: Vec<u8>,
        signature: Vec<u8>,
    },
    // ... add more schemes here
}
```

emmazzz commented:
> For multi-agents, we have the different authenticator AccountAuthenticator and we added the same variant to the authenticator as follows.
In the future, if we have a new scheme to support (e.g., ECDSA), I think we don't have to add a new variant "ECDSA" to both of authenticators. We just need to add "ECDSA" scheme to Scheme and reuse the LazyEvaluation variant.

This sounds good to me.

Naming wise, I don't like `ToBeDeserialized`, which is just a placeholder name I used in the doc. I like the idea of having the word "opaque" in the name. Maybe something like `OpaqueAuthScheme`? I don't have strong preference though.

imsungchoi-fb commented:
> > For multi-agents, we have the different authenticator AccountAuthenticator and we added the same variant to the authenticator as follows.
> > In the future, if we have a new scheme to support (e.g., ECDSA), I think we don't have to add a new variant "ECDSA" to both of authenticators. We just need to add "ECDSA" scheme to Scheme and reuse the LazyEvaluation variant.
> 
> This sounds good to me.
> 
> Naming wise, I don't like `ToBeDeserialized`, which is just a placeholder name I used in the doc. I like the idea of having the word "opaque" in the name. Maybe something like `OpaqueAuthScheme`? I don't have strong preference though.

Yeah I feel that `OpaqueAuthScheme` would be the best name I also came up with. 
@zekun000 If you don't have any concerns on the name, I will update the PR.

bmwill commented:
> Also, we would like to optimize the performance of deserialization by delaying the validation to happen at the verification stage, where transactions can be processed in parallel. This optimization involves storing raw bytes representing public keys and signatures in the transaction authenticator and deserializing only when necessary.

I just want to point out that this PR does not achieve this goal since you are requiring all users to change the way they construct transactions in order to "benefit" from delayed validation.

bmwill commented:
> With a more generic adapter interface, other instantiations of Diem can reuse the infrastructure, use OpaqueAuthScheme variant to implement their own authentication scheme and signature verification algorithm without having to change any Diem code.

I also want to push back on this claim that people will be able to implement their own auth schemes without having to change any diem code. This is not the case since you are defining a mapping between a u8 and a particular scheme, in diem specific code, If i wanted to add another scheme type I would still need to change diem code to introduce a mapping from u8 -> new scheme. If we really wanted to achieve this particular goal we might have a better chance if we introduced a trait to represent authenticators and making diem code generic across a type of authenticator (although that would require quite a bit of work).

I could be missing something though or misunderstanding the intent here. Have you tested that you can introduce a new authenticator type without changing diem's code (built on top of this PR)?

imsungchoi-fb commented:
> > Also, we would like to optimize the performance of deserialization by delaying the validation to happen at the verification stage, where transactions can be processed in parallel. This optimization involves storing raw bytes representing public keys and signatures in the transaction authenticator and deserializing only when necessary.
> 
> I just want to point out that this PR does not achieve this goal since you are requiring all users to change the way they construct transactions in order to "benefit" from delayed validation.

You're right. For the existing `Ed25519` or `MultiEd25519` variants, this PR does not make any performance improvement yet.  I think we should follow up in another PR to bring the benefit to the existing variants. With the description I was trying to explain why we introduced vec<u8> for public key and signature in `OpaqueAuthScheme`. I think the description is a little bit confusing. I will make it more clear.

imsungchoi-fb commented:
> > With a more generic adapter interface, other instantiations of Diem can reuse the infrastructure, use OpaqueAuthScheme variant to implement their own authentication scheme and signature verification algorithm without having to change any Diem code.
> 
> I also want to push back on this claim that people will be able to implement their own auth schemes without having to change any diem code. This is not the case since you are defining a mapping between a u8 and a particular scheme, in diem specific code, If i wanted to add another scheme type I would still need to change diem code to introduce a mapping from u8 -> new scheme. If we really wanted to achieve this particular goal we might have a better chance if we introduced a trait to represent authenticators and making diem code generic across a type of authenticator (although that would require quite a bit of work).
> 
> I could be missing something though or misunderstanding the intent here. Have you tested that you can introduce a new authenticator type without changing diem's code (built on top of this PR)?

@emmazzz would answer this question better. I will try my best. 
I think this `OpaqueAuthScheme` is a kind of stepping stone for [the change](https://github.com/diem/diem/pull/8773) emma is working on. 

How this PR change is related with emma's change is as follows. 

In emma's change, we introduce a new trait `VMAdapter`. 

```rust
pub trait VMAdapter {
   fn verify_signature(txn: &SignedTransaction) -> Result<()>;
   ... // and other customizations 
}
```

This SignedTransaction includes an authenticator with the variant `OpaqueAuthScheme` introduced in this PR. This way other instantiations of Diem can implement `VMAdapter` trait and use `OpaqueAuthScheme` variant for the signed transaction, and deserialize the keys and signatures into their favorite schemes in `verify_signature` function.

bmwill commented:
I just took a look at that PR and I agree with @sblackshear [first reaction](https://github.com/diem/diem/pull/8773#discussion_r674401433) that the VMAdapter trait should probably take in a transaction type defined by an associated type parameter, e.g.

```rust
pub trait VMAdapter {
   type Transaction: <Maybe some traits here>;
   fn verify_signature(txn: &Self::Transaction) -> Result<()>;
}
```

Specifically if your goal is to have people be able to implement their own auth schemes without having to change diem code, then the current approach we're taking necessitates that you must support ed25519 (since the Authenticator enum has that as one of the variants).

Also, to be clear, my biggest hesitation with the approach this PR is taking is that it is changing a public facing interface and is not just an implementation detail. I would also expect that when changing some critical public facing interface a DIP would be required (of course i'm not an expert in what requires a dip or not)

emmazzz commented:
> Specifically if your goal is to have people be able to implement their own auth schemes without having to change diem code, then the current approach we're taking necessitates that you must support ed25519 (since the Authenticator enum has that as one of the variants).

I'd say having a TransactionTrait is definitely one of the steps we can take down the way. It's not part of the current step since introducing it will incur a lot of refactoring and code changes. 

> Also, to be clear, my biggest hesitation with the approach this PR is taking is that it is changing a public facing interface and is not just an implementation detail. 

I understand the hesitation. Do we have a way to achieve similar performance gains without changing the public facing interface though? One way is to delay the deserialization of `SignedTransaction` altogether? Is it doable?

zekun000 commented:
we should separate performance gain from adding agility. we can get performance gain simply by changing the enum to
```
    Ed25519 {
        public_key: vec<u8>,
        signature: vec<u8>,
    },
```
and delay deserialization to verification stage. (It doesn't break compatibility I think)

This PR is more on addressing the agility problem.

bmwill commented:
> I understand the hesitation. Do we have a way to achieve similar performance gains without changing the public facing interface though? One way is to delay the deserialization of SignedTransaction altogether? Is it doable?

Agreed with @zekun000, this PR does nothing to solve performance issues since it necessitates the end user to change how they construct a transaction for those performance gains to take effect, any performance gains should ideally be done as an implementation detail, perhaps in a way Zekun showed above.

bmwill commented:
> I'd say having a TransactionTrait is definitely one of the steps we can take down the way. It's not part of the current step since introducing it will incur a lot of refactoring and code changes.

I haven't taken a very close look at your VMAdaptor PR, but given the VMAdaptor is new I don't know if it would be significantly more challenging to do this now (depending on how we go about making the change). Of course i've only started looking at this issue so i could very well be wrong.

bmwill commented:
Depending on how far we want to push for "agility" we could look towards doing something like https://github.com/bmwill/diem/commit/ac4c6a24582537ccc2cd3f6af35cb61c9143099f. This introduces a trait `TransactionAuthenticator` and makes `SignedTransaction` generic over `A: TransactionAuthenticator`. This will, at the end of the day, enable completely switching out the transaction authenticator to something else (maybe a type that doesn't even support ed25519) at the cost of adding generic type parameters in a number of places (I started this [here](https://github.com/bmwill/diem/commits/authenticator) but theres still a ton of spots that would need to be converted). Doing this, and maybe even abstracting a number of other concepts, would enable us to instantiate a blockchain very different from the existing DPN one we have today.

imsungchoi-fb commented:
> Depending on how far we want to push for "agility" we could look towards doing something like [bmwill@ac4c6a2](https://github.com/bmwill/diem/commit/ac4c6a24582537ccc2cd3f6af35cb61c9143099f). This introduces a trait `TransactionAuthenticator` and makes `SignedTransaction` generic over `A: TransactionAuthenticator`. This will, at the end of the day, enable completely switching out the transaction authenticator to something else (maybe a type that doesn't even support ed25519) at the cost of adding generic type parameters in a number of places (I started this [here](https://github.com/bmwill/diem/commits/authenticator) but theres still a ton of spots that would need to be converted). Doing this, and maybe even abstracting a number of other concepts, would enable us to instantiate a blockchain very different from the existing DPN one we have today.

I like the approach and I think it is more generic for others to instantiate a blockchain. I am willing to dig into this approach, but I am afraid that this would be overkill for the change we are making. 

Some questions on @emmazzz and @sblackshear 
* Any concerns/thought on this approach? 
* An alternative option between the current implementation and brandon's proposal? i.e., a stepping stone solution towards brandon's proposal

bmwill commented:
> I am afraid that this would be overkill for the change we are making.

You are absolutely correct, I'm most likely missing context for the intent behind this PR and there might be a simpler way to solve this depending on your goals.

emmazzz commented:
> Depending on how far we want to push for "agility" we could look towards doing something like bmwill@ac4c6a2. This introduces a trait TransactionAuthenticator and makes SignedTransaction generic over A: TransactionAuthenticator. This will, at the end of the day, enable completely switching out the transaction authenticator to something else (maybe a type that doesn't even support ed25519) at the cost of adding generic type parameters in a number of places (I started this here but theres still a ton of spots that would need to be converted). Doing this, and maybe even abstracting a number of other concepts, would enable us to instantiate a blockchain very different from the existing DPN one we have today.

I like this approach too! And it actually looks easier to implement than I expected. I actually don't mind doing this for agility and I think it aligns with what `VMAdapter` trait hopes to achieve. 

This way we can also separate the agility and performance conversations.

For performance, it seems that it's better to do the implementation specific approach @zekun000 originally proposed and what @bmwill proposed [here](https://github.com/diem/diem/pull/8950), which I think is the same essential idea.

xli commented on types/src/transaction/authenticator.rs:
```diff
@@ -240,6 +307,15 @@ pub enum Scheme {
+        match value {
+            0 => Scheme::Ed25519,
+            1 => Scheme::MultiEd25519,
+            _ => panic!("Unknown value: {}", value),
```
should not panic, this cause a client can crash server  by submitting a transaction with an invalid scheme number.

xli reviewed (commented):
Please add a test to json-rpc/integration-tests suite for fetching a transaction with new authenticator, and update doc https://github.com/diem/diem/blob/main/json-rpc/docs/type_transaction.md#type-transactiondata as necessary.

Also, I have same concern with @bmwill, I don't see a way you can actually add a new authenticator without changing Scheme definition or TransactionAuthenticator implementation on server side.
It would be great if you could add a test to show how this new OpaqueAuthScheme is used for the purpose of adding new authenticator.

imsungchoi-fb commented:
> > Depending on how far we want to push for "agility" we could look towards doing something like [bmwill/diem@ac4c6a2](https://github.com/bmwill/diem/commit/ac4c6a2). This introduces a trait TransactionAuthenticator and makes SignedTransaction generic over A: TransactionAuthenticator. This will, at the end of the day, enable completely switching out the transaction authenticator to something else (maybe a type that doesn't even support ed25519) at the cost of adding generic type parameters in a number of places (I started this here but theres still a ton of spots that would need to be converted). Doing this, and maybe even abstracting a number of other concepts, would enable us to instantiate a blockchain very different from the existing DPN one we have today.
> 
> I like this approach too! And it actually looks easier to implement than I expected. I actually don't mind doing this for agility and I think it aligns with what `VMAdapter` trait hopes to achieve.
> 
> This way we can also separate the agility and performance conversations.
> 
> For performance, it seems that it's better to do the implementation specific approach @zekun000 originally proposed and what @bmwill proposed [here](https://github.com/diem/diem/pull/8950), which I think is the same essential idea.

Okay I will try to make change the PR as guided in here [bmwill/diem@ac4c6a2]! Thanks for the feedback.

imsungchoi-fb commented:
> Please add a test to json-rpc/integration-tests suite for fetching a transaction with new authenticator, and update doc https://github.com/diem/diem/blob/main/json-rpc/docs/type_transaction.md#type-transactiondata as necessary.
> 
> Also, I have same concern with @bmwill, I don't see a way you can actually add a new authenticator without changing Scheme definition or TransactionAuthenticator implementation on server side.
> It would be great if you could add a test to show how this new OpaqueAuthScheme is used for the purpose of adding new authenticator.

Thanks for the feedback. 
I will add a new test in next change and update the doc. 
Also I will change the PR as guided in [bmwill/diem@ac4c6a2] which addresses your concerns.

imsungchoi-fb commented:
> > just small nit, it's probably better to have a type alias like DiemSignedTransaction = SignedTransaction.
> > or rename the SignedTransaction to something else and type alias SignedTransaction = GenericSignedTransaction.
> 
> Thanks for taking a look!
> 
> I guess that it is for easier refactoring in the future. It would be great if you elaborate more.
> I consider adding a type alias like
> 
> ```rust
> type DiemSignedTransaction = SignedTransaction<DiemTransactionAuthenticator>;
> ```
> 
> Do you think that we have to refactor codes to be relied on `DiemSignedTransaction` instead of `SignedTransaction<DiemTransactionAuthenticator>`?

Updated. 
Added two type aliases, `DiemSignedTransaction` and `DiemSignatureCheckedTransaction`.

xli commented on types/src/transaction/mod.rs:
```diff
@@ -453,6 +455,8 @@ impl WriteSetPayload {
     }
 }
 
+pub type DiemSignedTransaction = SignedTransaction<DiemTransactionAuthenticator>;
```
what is client code may see? A DiemSignedTransaction or SignedTransaction?

Current transaction builder generates type code for client applications, used name SignedTransaction.
If this change will cause generated client code to have DiemSignedTransaction, I think it's a bad idea to rename it.

xli replied:
More specifically, SignedTransaction is the part of interface to client applications, we should not break it.
